rust/kernel/str.rs

```diff
@@ -0,0 +1,225 @@
+    /// checks.
+    ///
+    /// # Safety
+    ///
+    /// `bytes` *must* end with a `NUL` byte, and should only have a single
+    /// `NUL` byte (or the string will be truncated).
+    #[inline]
+    pub const unsafe fn from_bytes_with_nul_unchecked(bytes: &[u8]) -> &CStr {
+        // Note: This can be done using pointer deref (which requires
+        //       const_raw_ptr_deref to be const) or transmute (which requires
+        //       const_transmute to be const) or ptr::from_raw_parts (which
+        //       requires ptr_metadata).
+        //       While none of them are current stable, it is very likely that
+        //       one of them will eventually be.
+        &*(bytes as *const [u8] as *const Self)
+    }
+
+    /// Returns a C pointer to the string.
+    #[inline]
+    pub const fn as_char_ptr(&self) -> *const c_types::c_char {
+        self.0.as_ptr() as _
+    }
+
+    /// Convert the string to a byte slice without the trailing 0 byte.
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0[..self.0.len() - 1]
+    }
+
+    /// Convert the string to a byte slice containing the trailing 0 byte.
+    #[inline]
+    pub const fn as_bytes_with_nul(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl AsRef<BStr> for CStr {
+    #[inline]
+    fn as_ref(&self) -> &BStr {
+        self.as_bytes()
+    }
+}
+
+impl Deref for CStr {
+    type Target = BStr;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        self.as_bytes()
+    }
+}
+
+impl Index<ops::RangeFrom<usize>> for CStr {
+    type Output = CStr;
+
+    #[inline]
+    fn index(&self, index: ops::RangeFrom<usize>) -> &Self::Output {
+        assert!(
+            index.start <= self.len(),
+            "range start index {} out of range for slice of length {}",
+            index.start,
+            self.len()
+        );
+        // SAFETY: We just checked the length.
+        unsafe { Self::from_bytes_with_nul_unchecked(&self.0[index.start..]) }
+    }
+}
+
+impl Index<ops::RangeFull> for CStr {
+    type Output = CStr;
+
+    #[inline]
+    fn index(&self, _index: ops::RangeFull) -> &Self::Output {
+        self
+    }
+}
+
+#[doc(hidden)]
+pub trait CStrIndex {}
+
+impl CStrIndex for usize {}
+impl CStrIndex for ops::Range<usize> {}
+impl CStrIndex for ops::RangeInclusive<usize> {}
+impl CStrIndex for ops::RangeToInclusive<usize> {}
+
+impl<Idx> Index<Idx> for CStr
+where
+    Idx: CStrIndex,
+    BStr: Index<Idx>,
+{
+    type Output = <BStr as Index<Idx>>::Output;
+
+    #[inline]
+    fn index(&self, index: Idx) -> &Self::Output {
+        &self.0[index]
+    }
+}
+
+/// Creates a new [`CStr`] from a string literal.
+///
+/// The string literal should not contain any `NUL` bytes.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// const MY_CSTR: &'static CStr = c_str!("My awesome CStr!");
+/// ```
+#[macro_export]
+macro_rules! c_str {
+    ($str:literal) => {{
+        // FIXME: Check that `$str` does not contain interior `NUL`.
```

Comment:
It doesn't look like we can have `for` loops in const functions, but the following version seems to work:

```
const fn has_null(str: &[u8]) -> bool {
    let mut i = 0;
    loop {
        if i >= str.len() {
            return false;
        }
        
        if str[i] == 0 {
            return true;
        }
        
        i += 1;
    }
}
```

Can we use it to remove this `FIXME`?

Reply:
Cool! So we don't actually proc macros to check for NULs. If I replace the `contains` call with while loops I could just use `from_bytes_with_nul` to do the check. I have used this trick elsewhere but somehow neglect it when doing this PR. 🤦